### PR TITLE
fix(mempool parity): Enforce comet / app-side mempool parity in CheckTx + integ. tests [BLO-584] [BLO-635]

### PR DESCRIPTION
**Pull request by nivasan1**
## In This PR
- I add a condition to **CheckTx** that removes txs from comet that are not app-side on `reCheckTx` 
- I add network-integration tests that test mempool parity when spammed w/ free-txs
- I add network-integration tests that test mempool parity when a bid invalidates a tx that would otherwise be valid
      - Feel free to leave comments / ask questions on `TestLanedMempoolSyncWithComet/bid Verify invalidates later tx Verify` as the edge-case is relatively involved

### CODE REVIEW

File: testutils/networksuite/networksuite.go [hunk 1]
```diff
@@ -27,13 +33,43 @@ import (
...
+			simtestutil.EmptyAppOptions{},
+			baseapp.SetPruning(pruningtypes.NewPruningOptionsFromString(val.GetAppConfig().Pruning)),
+			baseapp.SetMinGasPrices(val.GetAppConfig().MinGasPrices),
+			baseapp.SetChainID(chainID),
+		)
+	}
+
+	genBalance = sdk.NewCoin(sdk.DefaultBondDenom, math.NewInt(1000000000000000000))
+
+	txc client.TxConfig
```
Comment (Eric-Warehime) [inline testutils/networksuite/networksuite.go:54]: Very similar to what I pushed before in the SetupSuite function. Let's either use/update that stuff or move it up here and delete what's in that function. Doesn't make much difference to me where it gets set, but let's not have it in two places.

File: testutils/networksuite/account.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	return sdk.AccAddress(a.pk.PubKey().Address())
+}
+
+// PubKey returns the public-key of the account.
+func (a Account) PubKey() cryptotypes.PubKey {
+	return a.pk.PubKey()
+}
+
+// CreateTx creates and signs a transaction, from the given messages
+func (a Account) CreateTx(ctx context.Context, accNum, seq, gasLimit, fee, timeoutHeight uint64, msgs ...sdk.Msg) ([]byte, error) {
```
Comment (Eric-Warehime) [inline testutils/networksuite/account.go:38]: I pushed a similar function to chaintestutil https://github.com/skip-mev/chaintestutil/commit/7545b25770208ee7309ec55274416cc499419bf7

It seems like a function we can use across any package we want to test w/ that. Fine to leave this here and refactor it out later once we have a more stable set of chaintestutil capabilities implemented if you want to do that.

File: tests/integration/network/mempool_test.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Validators) == 0 {
+		return nil, nil
+	}
+
+	return sdk.ValAddressFromBech32(resp.Validators[0].OperatorAddress)
+}
```
Comment (Eric-Warehime) [inline tests/integration/network/mempool_test.go:170]: What's the difference between doing this and `s.NetworkSuite.Network.Validators[0].ValAddress`?

File: lanes/mev/check_tx.go [hunk 4]
```diff
@@ -64,17 +68,20 @@ type (
 	}
 )
 
-// NewCheckTxHandler constructs a new CheckTxHandler instance.
+// NewCheckTxHandler constructs a new CheckTxHandler instance. This method fails if the given LanedMempool does not have a lane
+// adhering to the MevLaneI interface
 func NewCheckTxHandler(
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply lanes/mev/check_tx.go:73]: Ahh, this is a great point actually. I'll do that

File: lanes/mev/check_tx.go [hunk 5]
```diff
@@ -121,6 +128,23 @@ func (handler *CheckTxHandler) CheckTx() CheckTx {
 			), nil
 		}
 
+		// if the mode is ReCheck and the app's mempool does not contain the given tx, we fail
+		// immediately, to purge the tx from the comet mempool.
+		if req.Type == cometabci.CheckTxType_Recheck && !handler.mempool.Contains(tx) {
+			handler.baseApp.Logger().Error(
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply lanes/mev/check_tx.go:134]: Great point, I j wanted to ability to introspect when this happened in normal operation, I'll move to Debug?

File: tests/integration/network/mempool_test.go (hunk 3 above)
Comment (nivasan1) [reply tests/integration/network/mempool_test.go:170]: Did not know we could do that lol

File: testutils/networksuite/account.go (hunk 2 above)
Comment (nivasan1) [reply testutils/networksuite/account.go:38]: I'll move to using chaintestutil

File: lanes/mev/check_tx.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Eric-Warehime) [reply lanes/mev/check_tx.go:134]: Yeah I guess debug or info? Seems like a somewhat normal reason for recheck tx to fail I guess is my point.

File: testutils/networksuite/account.go (hunk 2 above)
Comment (nivasan1) [reply testutils/networksuite/account.go:38]: On second thought, it looks like the helper there uses the val-keyring in the testsuite? I'd imagne that going forwards we want to be able to submit txs from accounts of our choice, and may want additional nuance in how we do things? Hence I'd prefer keeping thing like this, and refactoring out logic for tx-generation when we have a more educated opinion?

Comment (Eric-Warehime) [reply testutils/networksuite/account.go:38]: Seems like we're going to move this to chaintestutil via https://github.com/skip-mev/chaintestutil/pull/5/files

File: abci/checktx/mempl_parity_check_tx.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
+package checktx
```
Comment (aljo242) [inline abci/checktx/mempl_parity_check_tx.go:1]: let's rename to `mempool_parity_check_tx.go`

**Review by Eric-Warehime — APPROVED** [review]

File: lanes/mev/lane.go [hunk 7]
```diff
@@ -14,8 +14,6 @@ const (
 	LaneName = "mev"
 )
 
-var _ MEVLaneI = (*MEVLane)(nil)
-
```
[the remark this answers is not in the document]
Comment (nivasan1) [reply lanes/mev/lane.go:18]: MEVLaneI is an iface representing expectation of MEVLane dependency, so I moved to `mev_check_tx.go`

**Pull request merged**